File: samlang-ast/src/main/kotlin/samlang/ast/common/Literal.kt
package samlang.ast.common

/**
 * A collection of all supported literals.
 */
sealed class Literal {
    /**
     * A value to be pretty printed.
     */
    abstract val prettyPrintedValue: String

    /**
     * An int literal, like 42.
     */
    data class IntLiteral(val value: Long) : Literal() {
        override val prettyPrintedValue: String = value.toString()
    }

    /**
     * A string literal, like `"Answer to life, universe, and everything"`.
     */
    data class StringLiteral(val value: String) : Literal() {
        override val prettyPrintedValue: String = "\"$value\""
    }

    /**
     * A boolean literal, like `true` or `false`.
     */
    data class BoolLiteral(val value: Boolean) : Literal() {
        override val prettyPrintedValue: String = value.toString()
    }

    companion object {
        val TRUE: BoolLiteral = BoolLiteral(value = true)
        val FALSE: BoolLiteral = BoolLiteral(value = false)

        fun of(value: Long): IntLiteral = IntLiteral(value = value)

        fun of(value: String): StringLiteral = StringLiteral(value = value)
    }
}

File: typescript-rewrite/samlang-core/src/ast/common/Literal.ts
/** A collection of all supported literals. */

/** A boolean literal, like `true` or `false`.  */
export type BoolLiteral = { readonly type: 'BoolLiteral'; readonly value: boolean };
/** An int literal, like 42. */
export type IntLiteral = { readonly type: 'IntLiteral'; readonly value: bigint };

/** A string literal, like `"Answer to life, universe, and everything"`. */
export type StringLiteral = { readonly type: 'StringLiteral'; readonly value: string };
export type Literal = IntLiteral | StringLiteral | BoolLiteral;

export const TRUE: BoolLiteral = { type: 'BoolLiteral', value: true };
export const FALSE: BoolLiteral = { type: 'BoolLiteral', value: false };

export const intLiteralOf = (value: bigint): IntLiteral => ({
  type: 'IntLiteral',
  value,
});
export const stringLiteralOf = (value: string): StringLiteral => ({
  type: 'StringLiteral',
  value,
});

export const prettyPrintLiteral = (literal: Literal): string => {
  switch (literal.type) {
    case 'BoolLiteral':
    case 'IntLiteral':
      return String(literal.value);
    case 'StringLiteral':
      return `"${literal.value}"`;
  }
};
